sympy/solvers/solveset.py

```diff
@@ -2438,10 +2438,13 @@ def linear_coeffs(eq, *syms, **_kw):
     ...
     NonlinearError: nonlinear term encountered: x*(y + 1)
     """
+    from sympy.core.traversal import iterfreeargs
     d = defaultdict(list)
     eq = _sympify(eq)
     symset = set(syms)
-    has = eq.free_symbols & symset
+    if len(symset) != len(syms):
+        raise ValueError('duplicate symbols given')
+    has = set(iterfreeargs(eq)) & symset
     if not has:
```

Comment:
Shouldn't this just be `if eq.has_free(*symset)`?

Reply:
The set `has` is needed later, so instead of traversing twice, it is collected once.